state-chain/runtime/src/chainflip/vault_swaps.rs

```
@@ -108,6 +109,119 @@ pub fn bitcoin_vault_swap(
+					destination_asset,
+					ccm.message.to_vec(),
+					ccm.gas_budget,
+					cf_parameters,
+				)
+				.abi_encoded_payload())
+			} else {
+				Ok(cf_chains::evm::api::x_swap_native::XSwapNative::new(
+					destination_address,
+					destination_asset,
+					cf_parameters,
+				)
+				.abi_encoded_payload())
+			},
+		Asset::Flip | Asset::Usdc | Asset::Usdt | Asset::ArbUsdc => {
+			// Lookup Token addresses depending on the Chain
+			let source_token_address = match source_asset {
+				Asset::Flip | Asset::Usdc | Asset::Usdt =>
+					<EvmEnvironment as EvmEnvironmentProvider<Ethereum>>::token_address(
+						source_asset.try_into().expect("Only Ethereum asset is processed here"),
+					),
+				Asset::ArbUsdc =>
+					<EvmEnvironment as EvmEnvironmentProvider<Arbitrum>>::token_address(
+						cf_chains::assets::arb::Asset::ArbUsdc,
+					),
+				_ => unreachable!("Unreachable for non-Ethereum/Arbitrum assets"),
+			}
+			.ok_or(DispatchErrorWithMessage::from("Failed to look up EVM token address"))?;
+
+			if let Some(ccm) = channel_metadata {
+				Ok(cf_chains::evm::api::x_call_token::XCallToken::new(
+					destination_address,
+					destination_asset,
+					ccm.message.to_vec(),
+					ccm.gas_budget,
+					source_token_address,
+					amount,
+					cf_parameters,
+				)
+				.abi_encoded_payload())
+			} else {
+				Ok(cf_chains::evm::api::x_swap_token::XSwapToken::new(
+					destination_address,
+					destination_asset,
+					source_token_address,
+					amount,
+					cf_parameters,
+				)
+				.abi_encoded_payload())
+			}
+		},
+		_ => Err(DispatchErrorWithMessage::from(
+			"Only EVM chains should execute this branch of logic. This error should never happen",
+		)),
+	}?;
+
+	match source_asset.into() {
+		ForeignChain::Ethereum => Ok(VaultSwapDetails::Ethereum(EvmVaultSwapDetails {
+			calldata,
+			// Only return `amount` for native currently. 0 for Tokens
+			value: (source_asset == Asset::Eth).then_some(U256::from(amount)).unwrap_or_default(),
```

If I understand this code correctly, if the `U256::from(amount)` fails it will default to 0 too. If that's true, I think we should return an error in that scenario because something has gone wrong with the value and I'd rather error out.

---

Oh i see the confusion. 
the logic goes like this:
For Eth -> true -> evaluate(U256::from(amount)) -> Some(Amount) -> unwrap -> Amount
for Tokens -> false -> evaluate to None -> unwrap to default -> 0

So it's just a 1 liner way of saying: return 0 for tokens, Amount for Eth/ArbEth